src/mpid/ch4/generic/gpu/mpidig_gpu_utils.h

```diff
@@ -0,0 +1,183 @@
+/*
+ *  Copyright (C) by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ */
+#ifndef MPIDIG_GPU_UTIL_H_INCLUDED
+#define MPIDIG_GPU_UTIL_H_INCLUDED
+
+/* Define a threshold of high-density datatype */
+#define MPIDIG_GPU_DT_HIGH_DENSITY 0.5
+
+/* MPIDIG_gpu_alloc_stage_buf - allocate stage buffer and return pointer
+ * Stage buffer should be contiguous. */
+MPL_STATIC_INLINE_PREFIX void *MPIDIG_gpu_alloc_stage_buf(MPI_Aint count, MPI_Datatype datatype)
+{
+    char *tmp_buf = NULL;
+    size_t size;
+    MPIDI_Datatype_check_size(datatype, count, size);
+    MPL_gpu_malloc_host((void **) &tmp_buf, size);
+
+    return tmp_buf;
+}
+
+/* MPIDIG_gpu_free_stage_buf - free stage buffer */
+MPL_STATIC_INLINE_PREFIX void MPIDIG_gpu_free_stage_buf(void *buf)
+{
+    MPL_gpu_free_host(buf);
+}
+
+/* MPIDIG_gpu_stage_copy_gtoc - moving data from GPU buffer to Stage CPU buffer. */
+MPL_STATIC_INLINE_PREFIX void MPIDIG_gpu_stage_copy_gtoc(const void *sbuf, void *dbuf,
+                                                         MPI_Aint count, MPI_Datatype datatype)
+{
+    int dt_contig;
+    size_t size;
+    MPIR_Datatype *dt_ptr;
+    MPI_Aint true_lb;
+    yaksa_request_t request;
+    uintptr_t actual_pack_bytes;
+    int rc;
+
+    MPIDI_Datatype_get_info(count, datatype, dt_contig, size, dt_ptr, true_lb);
+    if (dt_contig) {
+        /* Stage buffer is contiguous, we can move directly. */
+        rc = yaksa_ipack(sbuf, (uintptr_t) size, YAKSA_TYPE__BYTE, 0,
+                         dbuf, (uintptr_t) size, &actual_pack_bytes, &request);
+        MPIR_Assert(rc == YAKSA_SUCCESS);
+        rc = yaksa_request_wait(request);
+        MPIR_Assert(rc == YAKSA_SUCCESS);
+    } else {
+        double density = ((double) (dt_ptr->size)) / ((double) (dt_ptr->extent));
```

Comment:
"Density" is really the average contiguous chunk length.  Maybe you can call it that instead of "density".  The formula to calculate that would be `dt_ptr->size / dt_ptr->max_contig_blocks`.  It's not related to the extent.  For example, if I have a contiguous array with a datatype traversing it in reverse order, extent will be the same as size, but each contiguous block is very small.